### PR TITLE
concepts/extend-kubernetes/api-extension: add 1.15 features

### PR DESCRIPTION
**Pull request by sttts**
Doc changes for:

* structural schema
* pruning
* defaulting
* conversion to beta
* publishing to beta

### CODE REVIEW

**Comment by sttts** [comment]
/assign @jpbetz @roycaihw @liggitt @caesarxuchao 

A first draft of the CRD doc updates. Ptal.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 1]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+  selfLink: /apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
+  uid: 9423255b-4600-11e7-af6a-28d2447dc82b
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
+The field `someRandomField` has been pruned.
+
+In `apiextensions/v1beta1`, pruning is disabled by default, i.e. `spec.preserveUnknownFields` defaults to `true. In `apiextensions/v1` no new CustomResourceDefinitions with `spec.preserveUnknownFields: true` will be allowed to be created.
```
Comment (tengqm) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: backtick after "true" is missing

Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: German Mac keyboard... not good for markdown.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 2]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+
+{{< feature-state state="beta" for_kubernetes_version="1.15" >}}
+
+CustomResources traditionally store arbitrary JSON (next to `apiVersion`, `kind` and `metadata`, which is validated by the API server implicitly). With [OpenAPI v3.0 validation](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) a schema can be specified, which is validated during creation and updates, compare below for details and limits of such a schema.
+
+With `apiextensions.k8s.io/v1` the definition of a structural schema will be mandatory for CustomResourceDefinitions, while in `v1beta1` this is still optional.
+
+A structural schema is an [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) which:
+
+1. specifies a non-empty type (via `type` in OpenAPI) for the root, for each specified field of an object node (via `properties` or `additionalProperties in OpenAPI) and for each item in an array node (via `items` in OpenAPI), with the exception of:
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 1 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 3]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+
+With `apiextensions.k8s.io/v1` the definition of a structural schema will be mandatory for CustomResourceDefinitions, while in `v1beta1` this is still optional.
+
+A structural schema is an [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) which:
+
+1. specifies a non-empty type (via `type` in OpenAPI) for the root, for each specified field of an object node (via `properties` or `additionalProperties` in OpenAPI) and for each item in an array node (via `items` in OpenAPI), with the exception of:
+   * a node with `x-kubernetes-int-or-string: true`
+   * a node with `x-kubernetes-preserve-unknown-fields: true`
+2. specifies each field in an object and each item in an array which is specified within an `allOf`, `anyOf`, `oneOf` or `not` also outside of those.
```
Comment (ironcladlou) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: I'm not sure yet how I would state it more clearly, but I had to re-read this point and cross-reference it with the examples multiple times to understand the rule being described. I'm not sure whether that's a reflection of quality of the sentence or my own ignorance of the subject matter.

Is it true that property names declared in `allOf` (etc.) must also be declared in the root `properties`?

Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]:
```yaml
allOf:
- properties:
    foo: X
```

means that you have to specify `foo` at the root as well:

```yaml
properties:
  foo: Y
allOf:
- properties:
    foo: X
```

Y can differ form X, i.e. only contain the type.

If you have a better way to formulate that, let me know.

Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: And this is transitive for any depth and including arrays, e.g.

```yaml
allOf:
- items:
    properties:
      foo: X
```

means that you have to specify `[].foo` at the root as well:

```yaml
items:
  properties:
    foo: Y
allOf:
- items:
    properties:
      foo: X
```

Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Luckily you get all of this for free with a generator like kubebuilder's controller-gen because it is just the way it work by traversing the API types.

Comment (roycaihw) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: I also had to re-read this a couple of times (it became concise once you got the idea). Perhaps break the sentence into smaller pieces, or move the subject around?

maybe something (ugly) like
> for each field in an object and each item in an array which is specified within a logical junctor (`allOf`, `anyOf`, `oneOf` or `not`), the schema also specifies the field/item outside of those logical junctors"

or we can use these examples as well

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 4]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+### RawExtension
+
+RawExtensions (as in `runtime.RawExtension` defined in [k8s.io/apimachinery](https://github.com/kubernetes/apimachinery/blob/03ac7a9ade429d715a1a46ceaa3724c18ebae54f/pkg/runtime/types.go#L94)) holds complete Kubernetes objects, i.e. with `apiVersion` and `kind` fields.
+
+It is possible specify those embedded objects (both completely without constraints or partially specified) by setting `x-kubernetes-embedded-object: true`. For example:
+
+```yaml
+type: object
+properties:
+  embedded:
```
Comment (roycaihw) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: nit: somehow I had the first impression that the name `embedded` is special... can we rename it to something like `foo` or `embeddedFoo`?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 5]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
+The field `someRandomField` has been pruned.
+
+In `apiextensions/v1beta1`, pruning is disabled by default, i.e. `spec.preserveUnknownFields` defaults to `true`. In `apiextensions/v1` no new CustomResourceDefinitions with `spec.preserveUnknownFields: true` will be allowed to be created.
+
+### Controlling pruning
```
Comment (roycaihw) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: nit: this is not specific to this PR

I wish the table of content looks like
```
- Specifying a structural schema
  - structural schema
    # group these three together
  - customizing structural schema with openapi extensions
    - Controlling pruning
    - IntOrString
    - RawExtension
- Advanced topics
  - Validation
  - Defaulting
    # I feel pruning doesn't have to be under structural schema
  - Pruning
```
but currently it only shows top level: https://deploy-preview-14583--kubernetes-io-vnext-staging.netlify.com/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#advanced-topics
```
- Specifying a structural schema
- Advanced topics
```

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 3 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: added the examples and took @roycaihw sentence.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 6]
```diff
@@ -347,57 +641,111 @@ kubectl apply -f my-crontab.yaml
...
 ```
 
-Custom resource validation schema will be converted to OpenAPI v2 schema, and
+Note that defaulting happens on the object
+ 
+* in the request to the API server using the request version defaults
+* when reading from etcd using the storage version defaults
+* after mutating admission plugins with non-empty patches using the admission webhook object version defaults.
+
+E.g. adding a default to the schema will automatically result in the defaults being applied from getting or listing objects, e.g. via `kubectl get`. Note though that defaults are not automatically written to the persisted objects.
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 7]
```diff
@@ -569,7 +917,7 @@ the status replica value in the `/scale` subresource will default to 0.
 
   - It is an optional value.
   - It must be set to work with HPA.
-  - Only JSONPaths under `.status` and with the dot notation are allowed.
+  - Only JSONPaths under `.status` and `.spec` and with the dot notation are allowed.
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: done

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 4 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: done

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 8]
```diff
@@ -347,57 +641,111 @@ kubectl apply -f my-crontab.yaml
...
 
-{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+{{< feature-state state="alpha" for_kubernetes_version="1.15" >}}
 
-Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
-of [OpenAPI v2 spec](/docs/concepts/overview/kubernetes-api/#openapi-and-swagger-definitions) from
-Kubernetes API server.
+{{< note >}}
+Defaulting is available as alpha since 1.15. The
+`CustomResourceDefaulting` feature must be enabled, which is the case automatically for many clusters for beta features. Please refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 5 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: I place pruning more at the top because this won't be opt-in anymore in 1.0.

**Comment by sttts** [comment]
@roycaihw addressed your comments.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 9]
```diff
@@ -178,6 +192,273 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+RawExtensions (as in `runtime.RawExtension` defined in [k8s.io/apimachinery](https://github.com/kubernetes/apimachinery/blob/03ac7a9ade429d715a1a46ceaa3724c18ebae54f/pkg/runtime/types.go#L94)) holds complete Kubernetes objects, i.e. with `apiVersion` and `kind` fields.
+
+It is possible specify those embedded objects (both completely without constraints or partially specified) by setting `x-kubernetes-embedded-object: true`. For example:
+
+```yaml
+type: object
+properties:
+  embedded:
+    x-kubernetes-embedded-object: true
+    x-kubernetes-preserve-unknown-fields: true
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: clarified

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 10]
```diff
@@ -231,17 +512,28 @@ Validation of custom objects is possible via
 [OpenAPI v3 schema](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject) or [validatingadmissionwebhook](/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook).
 Additionally, the following restrictions are applied to the schema:
 
-- The fields `default`, `nullable`, `discriminator`, `readOnly`, `writeOnly`, `xml`,
-`deprecated` and `$ref` cannot be set.
+- These fields cannot be set:
+  - `default` (if the `CustomResourceDefaulting` feature gate on the [kube-apiserver](/docs/admin/kube-apiserver) is off),
```
[the remark this answers is not in the document]
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: :+1:

fixed

**Comment by sttts** [comment]
@tengqm do you we need anybody from the docs team to review?

**Comment by makoscafee** [comment]
Hi @sttts, @tengqm  any assistance needed for this to be approved and merged? We are looking to have all PR for release 1.15 merged by June 10th. Let me know if there is anything we can help.

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 11]
```diff
@@ -178,6 +192,313 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+* [Validation Schema Defaulting](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#defaulting)
+* [Pruning](#preserving-unknown-fields)
+
+and possibly more features in the future.
+
+### Pruning versus preserving unknown fields
+
+{{< feature-state state="beta" for_kubernetes_version="1.15" >}}
+
+CustomResourceDefinitions traditionally store any (possibly validated) JSON as is in etcd. This means that unspecified fields (if there is a [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) at all) will be persisted. This is in contrast to native Kubernetes resources like e.g. a pod. If the user specifies a field `spec.randomField`, this will be accepted by the API server HTTPS endpoint.
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Isn't `If the user specifies a field `spec.randomField`, this will be accepted by the API server HTTPS endpoint.` missing the crucial detail "... but will be dropped before being persisted in etcd."?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 12]
```diff
@@ -178,6 +192,313 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+spec:
+  cronSpec: "* * * * */5"
+  image: my-awesome-cron-image
+  someRandomField: 42
+```
+
+and create it:
+
+```shell
+kubectl create -f my-crontab.yaml -o yaml
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: you actually have to add `--validate=false` to skip client-side validation complaining about the field, right?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 13]
```diff
@@ -231,17 +552,31 @@ Validation of custom objects is possible via
...
+  - `discriminator`, 
+  - `id`,
+  - `patternProperties`,
+  - `readOnly`, 
+  - `writeOnly`,
+  - `xml`,
+  - `$ref`.
 - The field `uniqueItems` cannot be set to true.
 - The field `additionalProperties` cannot be set to false.
+- The field `additionalProperties` is mutual exclusive with `properties`.
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: typo: mutually

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 14]
```diff
@@ -347,57 +684,116 @@ kubectl apply -f my-crontab.yaml
...
 
-{{< feature-state state="alpha" for_kubernetes_version="1.14" >}}
+{{< feature-state state="alpha" for_kubernetes_version="1.15" >}}
 
-Starting with Kubernetes 1.14, [custom resource validation schema](#validation) can be published as part
-of [OpenAPI v2 spec](/docs/concepts/overview/kubernetes-api/#openapi-and-swagger-definitions) from
-Kubernetes API server.
+{{< note >}}
+Defaulting is available as alpha since 1.15. The
+`CustomResourceDefaulting` feature must be enabled, which is the case automatically for many clusters for beta features. Please refer to the [feature gate](/docs/reference/command-line-tools-reference/feature-gates/) documentation for more information.
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: strike "which is the case automatically for many clusters for beta features" until this is a beta feature

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 15]
```diff
@@ -347,57 +684,116 @@ kubectl apply -f my-crontab.yaml
...
+  replaces: 1
+```
+
+Note that defaulting happens on the object
+ 
+* in the request to the API server using the request version defaults
+* when reading from etcd using the storage version defaults
+* after mutating admission plugins with non-empty patches using the admission webhook object version defaults.
+
+Hence, adding a default to the schema will automatically result in the defaults being applied from getting or listing objects, e.g. via `kubectl get`. Note though that defaults are not automatically written to the persisted objects.
```
Comment (liggitt) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: might be clearer as "Note that defaults applied when reading data from etcd are not automatically written back to etcd; an update request via the API is required to persist those defaults back into etcd."

**Comment by liggitt** [comment]
a few small comments, but lgtm overall

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 11 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 12 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: done

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 13 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 16]
```diff
@@ -178,6 +192,313 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+  selfLink: /apis/stable.example.com/v1/namespaces/default/crontabs/my-new-cron-object
+  uid: 9423255b-4600-11e7-af6a-28d2447dc82b
+spec:
+  cronSpec: '* * * * */5'
+  image: my-awesome-cron-image
+```
+
+The field `someRandomField` has been pruned.
+
+In `apiextensions/v1beta1`, pruning is disabled by default, i.e. `spec.preserveUnknownFields` defaults to `true`. In `apiextensions/v1` no new CustomResourceDefinitions with `spec.preserveUnknownFields: true` will be allowed to be created.
```
Comment (roycaihw) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: nit: s/`apiextensions/v...`/`apiextensions.k8s.io/v...`

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 14 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 15 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 16 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 17]
```diff
@@ -178,6 +192,315 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
 
 If you later recreate the same CustomResourceDefinition, it will start out empty.
 
+## Specifying a structural schema
+
+{{< feature-state state="beta" for_kubernetes_version="1.15" >}}
+
+CustomResources traditionally store arbitrary JSON (next to `apiVersion`, `kind` and `metadata`, which is validated by the API server implicitly). With [OpenAPI v3.0 validation](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) a schema can be specified, which is validated during creation and updates, compare below for details and limits of such a schema.
```
Comment (jpbetz) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Should we say JSON here? It's stored that way, but from the user perspective, it can be JSON or YAML..

Comment (jpbetz) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: s/traditionally/"early versions of CustomResources"?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 18]
```diff
@@ -178,6 +192,315 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+
+1. specifies a non-empty type (via `type` in OpenAPI) for the root, for each specified field of an object node (via `properties` or `additionalProperties` in OpenAPI) and for each item in an array node (via `items` in OpenAPI), with the exception of:
+   * a node with `x-kubernetes-int-or-string: true`
+   * a node with `x-kubernetes-preserve-unknown-fields: true`
+2. for each each field in an object and each item in an array which is specified within an `allOf`, `anyOf`, `oneOf` or `not`, the schema also specifies the field/item outside of those logical junctors (compare example 1 and 2).
+3. does not set `description`, `type`, `default`, `additionProperties`, `nullable` within an `allOf`, `anyOf`, `oneOf` or `not`, with the exception of the two pattern for `x-kubernetes-int-or-string: true` (see below).
+4. if `metadata` is specified, then only restrictions on `metadata.name` and `metadata.generateName` are allowed. 
+
+
+Example 1:
```
Comment (jpbetz) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: State upfront that this example is invalid? E.g. "Non-structural example"?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 19]
```diff
@@ -178,6 +192,315 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+
+With `apiextensions.k8s.io/v1` the definition of a structural schema will be mandatory for CustomResourceDefinitions, while in `v1beta1` this is still optional.
+
+A structural schema is an [OpenAPI v3.0 validation schema](/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/#validation) which:
+
+1. specifies a non-empty type (via `type` in OpenAPI) for the root, for each specified field of an object node (via `properties` or `additionalProperties` in OpenAPI) and for each item in an array node (via `items` in OpenAPI), with the exception of:
+   * a node with `x-kubernetes-int-or-string: true`
+   * a node with `x-kubernetes-preserve-unknown-fields: true`
+2. for each each field in an object and each item in an array which is specified within an `allOf`, `anyOf`, `oneOf` or `not`, the schema also specifies the field/item outside of those logical junctors (compare example 1 and 2).
```
Comment (jpbetz) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: nit: If we say "those logical junctors" also say "which is specified within one of the logical junctors.. " to help readers connect the dots?

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 17 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: everything up to 1.15 and everything not setting `spec.preserveUnknownFields: false`

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 18 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: done

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 19 above)
Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: fixed

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md [hunk 20]
```diff
@@ -178,6 +192,315 @@ Error from server (NotFound): Unable to list {"stable.example.com" "v1" "crontab
...
+properties:
+  foo:
+    ...
+allOf:
+- properties:
+    foo:
+      ...
+```
+
+Example 2:
```
Comment (jpbetz) [inline content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Non-Structural Example here and below with "Example 3" as well?

Comment (sttts) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: yep

File: content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md (hunk 17 above)
Comment (jpbetz) [reply content/en/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions.md]: Hm, that might be too wordy. "traditionally" might be okay..

**Comment by jimangel** [comment]
There's a fair amount of using the word "will" in the past tense. The style guide outlines that docs should be in present tense: https://kubernetes.io/docs/contribute/style/style-guide/#use-present-tense

However, for the sake of timing, the content is good and I would be willing to merge, as-is, and itterate on top.

/cc @cody-clark

**Comment by sttts** [comment]
Replaces the "will"s with present tense where it makes sense.

**Comment by sttts** [comment]
@jimangel can you retag?
